src/file_data.rs

```diff
@@ -0,0 +1,68 @@
+use std::fs::File;
+use std::path::{Path, PathBuf};
+
+use serde::Deserialize;
+
+use crate::data::{
+    DataSource, DataSourceInfo, EntryID, SlotMetaTile, SlotTile, SummaryTile, TileID,
+};
+use crate::http::schema::TileRequestRef;
+
+pub struct FileDataSource {
+    pub basedir: PathBuf,
+}
+
+impl FileDataSource {
+    pub fn new(basedir: impl AsRef<Path>) -> Self {
+        Self {
+            basedir: basedir.as_ref().to_owned(),
+        }
+    }
+
+    fn read_file<T>(&self, path: impl AsRef<Path>) -> T
+    where
+        T: for<'a> Deserialize<'a>,
+    {
+        let f = File::open(path).expect("opening file failed");
+        let f = zstd::Decoder::new(f).expect("zstd decompression failed");
+        let result = ciborium::from_reader(f).expect("cbor decoding failed");
+        result
+    }
+}
+
+impl DataSource for FileDataSource {
+    fn fetch_info(&self) -> DataSourceInfo {
+        let mut path = self.basedir.clone();
+        path.push("info");
```

Comment:
could `mut` be avoided here (and elsewhere)? 
```suggestion
        let path = self.basedir.join("info");
```

That said, I was a little confused by the docs for [`Path.join`](https://doc.rust-lang.org/stable/std/path/struct.Path.html#method.join). The clearly show an example of passing a string to `join` but I am not sure how to correlate that to the given `join<P: AsRef<Path>>(&self, path: P)` signature there. I guess it is to do with [this](https://doc.rust-lang.org/stable/src/std/path.rs.html#1737)?

Reply:
Thanks. I was looking at the `PathBuf` docs but somehow missed `join`.

I think the reason why strings work is because of the `impl` for `AsRef<Path>` on `String`:

https://doc.rust-lang.org/std/convert/trait.AsRef.html#impl-AsRef%3CPath%3E-for-String

(As much as I appreciate the flexibility of Rust, figuring out what traits are available is one of my ongoing annoyances with the language. Oftentimes something will be possible because trait X is implemented on type Y, but if you don't know about X (or that X is related to the feature you want, or that X is implemented for Y), then the resulting code can be very opaque.)